src/config.rs

```diff
@@ -59,37 +86,98 @@ impl fmt::Display for Device {
 }
 
 pub fn read_all_devices(root: &Path) -> Result<Vec<Device>> {
-    let memtotal_mb = get_total_memory_kb(&root)? as f64 / 1024.;
-    Result::from_iter(
-        read_devices(root)?
-            .map(|(sn, s)| parse_device(sn, s, memtotal_mb))
-            .map(Result::transpose)
-            .flatten(),
-    )
+    let memtotal_mb = get_total_memory_kb(&root)? / 1024;
+
+    let devices: Vec<Device> = read_devices(root, memtotal_mb)?
+        .into_iter()
+        .filter(|(_, dev)| dev.disksize > 0)
+        .map(|(_, dev)| dev)
+        .collect();
+
+    Ok(devices)
 }
 
-fn read_devices(root: &Path) -> Result<SectionIntoIter> {
-    let path = root.join("etc/systemd/zram-generator.conf");
-    if !path.exists() {
+fn read_devices(root: &Path, memtotal_mb: u64) -> Result<HashMap<String, Device>> {
+    let fragments = locate_fragments(root);
+
+    if fragments.is_empty() {
         println!("No configuration file found.");
-        return Ok(Ini::new().into_iter());
     }
 
-    Ok(Ini::load_from_file(&path)
-        .with_context(|| format!("Failed to read configuration from {}", path.display(),))?
-        .into_iter())
+    let mut devices: HashMap<String, Device> = HashMap::new();
+
+    for (_, path) in fragments {
+        // eprintln!("{}", path.display());
+
+        let ini = Ini::load_from_file(&path).unwrap();
+
+        for (sname, props) in ini.iter() {
+            let sname = match sname {
+                None => {
+                    eprintln!(
+                        "{:?}: ignoring settings outside of section: {:?}",
+                        path, props
+                    );
+                    continue;
+                }
+                Some(sname) => sname.to_string(),
+            };
+
+            if !sname.starts_with("zram") {
+                println!("Ignoring section \"{}\"", sname);
+                continue;
+            }
+
+            let dev = devices
+                .entry(sname.clone())
+                .or_insert_with(|| Device::new(sname));
+
+            for (k, v) in props.iter() {
+                parse_line(dev, k, v)?;
+            }
+        }
+    }
+
+    for (_, dev) in &mut devices {
+        dev.set_disksize_if_enabled(memtotal_mb);
+    }
+
+    Ok(devices)
+}
+
+fn locate_fragments(root: &Path) -> BTreeMap<String, PathBuf> {
+    let base_dirs = vec![
+        String::from(root.join("usr/lib").to_str().unwrap()),
+        String::from(root.join("usr/local/lib").to_str().unwrap()),
+        String::from(root.join("etc").to_str().unwrap()),
+        String::from(root.join("run").to_str().unwrap()), // We look at /run to allow temporary overriding
+                                                          // of configuration. There is no expectation of
+                                                          // programatic creation of config there.
+    ];
+
+    let cfg = FragmentScanner::new(
+        base_dirs.clone(),
```

Comment:
This does four PathBuf allocations, then four String allocations twice, then the vector twice.

The liboverdrop API is quite limiting here, but `base_dirs` can be a `[PathBuf; 4]` (`[root.join("usr/lib"), ...]`), then this argument can be `base_dirs.iter().cloned().map(|pb| pb.into_os_string().into_string().unwrap()).collect()`, which brings it down to four PathBufs, twice, then the vector once.

A comment or `expect()` arg so as to why the unwrap is fine (root is just / in normal operation and otherwise controlled in testing) would be nice as well.

Reply:
We are free to adjust the liboverdrop API, it isn't stellar about this right now. Signature suggestions (and PRs) welcome.

Reply:
Hmm, but does it matter? 4 allocations like this are completely negligible, and we do it just once.

Reply:
It might not matter for the purposes of performance, but it's very very wordy and it was hard to decipher what the code was *trying* to do. Consider this implementation instead, which does away with most of the fluff (fewer allocations are a bonus) and is easier to figure out:
```rust
fn locate_fragments(root: &Path) -> BTreeMap<String, PathBuf> {
    let mut base_dirs = [
        root.join("usr/lib"),
        root.join("usr/local/lib"),
        root.join("etc"),
        root.join("run"), // We look at /run to allow temporary overriding
                          // of configuration. There is no expectation of
                          // programatic creation of config there.
    ];

    let cfg = FragmentScanner::new(
        base_dirs.iter().cloned().map(|pb| pb.into_os_string().into_string().unwrap()).collect(),
        "systemd/zram-generator.conf.d",
        true,
        vec![String::from("conf")],
    );

    let mut fragments = cfg.scan();

    for path in base_dirs.iter_mut().rev() {
        path.push("systemd/zram-generator.conf");
        if path.exists() {
            fragments.insert(String::new(), path); // The empty string shall sort earliest
            break;
        }
    }

    fragments
}
```

Reply:
Can you submit this as a separate PR please?

Reply:
Always: #34